cpg_pipes/providers/inputs.py

```diff
@@ -0,0 +1,417 @@
+"""
+Abstract provider if input data source.
+"""
+
+import csv
+import logging
+from abc import ABC, abstractmethod
+from enum import Enum
+
+from ..utils import exists
+from ..types import FastqPair, CramPath, AlignmentInput
+from ..targets import Cohort, Dataset, Sex, SequencingType, PedigreeInfo
+from .. import Path
+
+logger = logging.getLogger(__file__)
+
+
+class InputProviderError(Exception):
+    """
+    Exception thrown when there is something wrong happened parsing
+    inputs.
+    """
+
+    pass
+
+
+class InputProvider(ABC):
+    """
+    Abstract class for implementing inputs source.
+    """
+
+    def __init__(self, check_files: bool = True):
+        self.check_files = check_files
+
+    def populate_cohort(
+        self,
+        cohort: Cohort,
+        dataset_names: list[str] | None = None,
+        skip_samples: list[str] | None = None,
+        only_samples: list[str] | None = None,
+        skip_datasets: list[str] | None = None,
+        ped_files: list[Path] | None = None,
+    ) -> Cohort:
+        """
+        Add datasets in the cohort. There exists only one cohort for
+        the pipeline run.
+        """
+        if dataset_names:
+            # Specific datasets requested, so initialising them in advance.
+            for ds_name in dataset_names:
+                if skip_datasets and ds_name in skip_datasets:
+                    logger.info(f'Requested to skipping dataset {ds_name}')
+                    continue
+                dataset = cohort.add_dataset(ds_name)
+                entries = self._get_entries(
+                    dataset,
+                    skip_samples=skip_samples,
+                    only_samples=only_samples,
+                    skip_datasets=skip_datasets,
+                )
+                for entry in entries:
+                    self._add_sample(dataset, entry)
+
+        else:
+            # We don't know dataset names in advance, so getting all entries.
+            entries = self._get_entries(
+                skip_samples=skip_samples,
+                only_samples=only_samples,
+                skip_datasets=skip_datasets,
+            )
+            for entry in entries:
+                ds_name = self.get_dataset_name(entry) or cohort.analysis_dataset.name
+                dataset = cohort.add_dataset(ds_name)
+                self._add_sample(dataset, entry)
+
+        self.populate_alignment_inputs(cohort)
+        self.populate_analysis(cohort)
+        self.populate_participants(cohort)
+        self.populate_pedigree(cohort)
+        if ped_files:
+            self.populate_pedigree_from_ped_files(cohort, ped_files)
+
+        return cohort
+
+    @abstractmethod
+    def get_entries(
+        self,
+        dataset: Dataset | None = None,
+    ) -> list[dict]:
+        """
+        Overide this method to get a list of data entries (dicts).
+        If dataset is not missing, it should be specific to a dataset.
```

Comment:
Not sure what's meant by this?

Reply:
If `dataset` parameter is provided, the method must return entries for this dataset only. I'll rename the parameter to `limit_to_dataset` to make it clearer